### PR TITLE
fix: pageserver_ondisk_layers metric

### PR DESCRIPTION
**Pull request by koivunej**
Current pageserver_ondisk_layers is wrong, see issue. To fix that, we need `trait PersistentLayer` in LayerMap. Solve this need by introducing a wrapper `MeasuredLayerMap` where the metrics are updated. Additionally an existing test which does evictions is modified to check this fixed metric (couldn't find an existing test case and nothing broke).

Additionally has three unrelated fixes.

Cc: #3705 (separate from adding a metric for remote layers)

### CODE REVIEW

**Comment by koivunej** [comment]
Need to check if we have any test for this metric, if not, fix that and then last split the metric up.

**Comment by koivunej** [comment]
> deref-abuse
Did a poll internally, no one is horrified.

**Comment by koivunej** [comment]
merge-allure-reports failed (debug)[https://github.com/neondatabase/neon/actions/runs/4374610146/jobs/7654510396]?

```
+ aws s3 cp --only-show-errors s3://neon-github-public-dev/reports/pr-3775/lock.txt ./lock.txt
fatal error: An error occurred (404) when calling the HeadObject operation: Key "reports/pr-3775/lock.txt" does not exist
```

**Review by problame — COMMENTED** [review]
Can we move the `resident_physical_size` gauge into the `MeasuredLayerMap` as well?

(If you still want to attempt a Collector-based approach, that should be a separate PR)

File: pageserver/src/tenant/layer_map.rs [hunk 1]
```diff
@@ -340,12 +333,15 @@ where
...
+                .l0_delta_layers
+                .iter()
+                .position(|slot| Self::compare_arced_layers(slot, expected));
+
+            if pos.is_none() {
+                return Ok(Replacement::NotFound);
+            }
+            pos
         } else {
             None
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/layer_map.rs:346]: In #3805.

**Comment by koivunej** [comment]
It's non-trivial to move all affected metrics in (I wasn't expecting there to many other) so drafting while waiting for time to look at this.

**Comment by koivunej** [comment]
Mentioned for generation/version number which would be handy in number of cases.

**Comment by koivunej** [comment]
Sad to leave this out because of scope creep.

**Pull request closed without merging**